pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -73,7 +73,7 @@ def __init__(
         self._has_spawned_children = False
         self.task_idx = None
         self.node_rank = 0
-        self.num_processes = len(parallel_devices)
+        self.num_processes = len(parallel_devices) if parallel_devices is not None else parallel_devices
```

Comment:
if parallel_devices can be None I think we need to change the root_devices access property
because in ddp root_device = parallel_devices[local_rank]

Reply:
They can't be none. They are optional for constructor, since these plugins will be provided by the user, but before calling anything else, they will be set in the accelerator_connector.